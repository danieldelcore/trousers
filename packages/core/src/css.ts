import hash from '@trousers/hash';

import { Definition, CSSObject } from './types';
import themify from './themify';

export interface CollectorReturn {
    _get: () => Definition[];
    modifier: (id: string, styles: CSSObject) => CollectorReturn;
    global: (styles: CSSObject) => CollectorReturn;
    theme: (theme: Record<string, any>) => CollectorReturn;
}

function css(styles: CSSObject): CollectorReturn;
function css(
    idOrStyle: string | CSSObject,
    styles?: CSSObject,
): CollectorReturn;
function css(
    idOrStyle: string | CSSObject,
    styles?: CSSObject,
): CollectorReturn {
    let id = typeof idOrStyle === 'string' ? idOrStyle : '';
    let styleObject = (typeof idOrStyle === 'string'
        ? styles
        : idOrStyle) as CSSObject;

    const elementId = `${id}-${hash(JSON.stringify(styleObject))}`;
    const styleMap: Definition[] = [
        {
            id,
            type: 'element',
            className: `.${elementId}`,
            styles: styleObject,
        },
    ];

    const self: CollectorReturn = {
        _get: () => styleMap,
        modifier: (modifierId, modifierStyles) => {
            styleMap.push({
                id: modifierId,
                type: 'modifier',
                className: `.${elementId}--${modifierId}-${hash(
                    JSON.stringify(modifierStyles),
                )}`,
                styles: modifierStyles,
            });

            return self;
        },
        global: globalStyles => {
            styleMap.push({
                id: `global-${id}-${hash(JSON.stringify(globalStyles))}`,
                type: 'global',
                className: '',
                styles: globalStyles,
            });

            return self;
        },
        theme: theme => {
            const themeHash = hash(JSON.stringify(theme));
            styleMap.push({
                id: `theme-${id}-${themeHash}`,
                type: 'theme',
                className: `.theme-${id}-${themeHash}`,
                styles: themify(theme),
            });

            return self;
        },
    };

    return self;
}

export type Collector = typeof css;

export default css;
