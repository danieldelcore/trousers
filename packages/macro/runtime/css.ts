interface Definition {
    type: 'element' | 'modifier' | 'global' | 'theme';
    id: string;
    styles: string;
}

interface CollectorReturn {
    modifier: (id: string, styles: string) => CollectorReturn;
    global: (id: string, styles: string) => CollectorReturn;
    theme: (id: string, styles: string) => CollectorReturn;
    _get: () => Definition[];
}

function css(elementId: string, elementStyles: string) {
    const styleMap: Definition[] = [
        {
            type: 'element',
            id: elementId,
            styles: elementStyles,
        },
    ];

    const self: CollectorReturn = {
        _get: () => styleMap,
        modifier: (modifierId, modifierStyles) => {
            styleMap.push({
                type: 'modifier',
                id: modifierId,
                styles: modifierStyles,
            });

            return self;
        },
        global: (globalId, globalStyles) => {
            styleMap.push({
                type: 'global',
                id: globalId,
                styles: globalStyles,
            });

            return self;
        },
        theme: (themeId, themeStyles) => {
            styleMap.push({
                type: 'theme',
                id: themeId,
                styles: themeStyles,
            });

            return self;
        },
    };

    return self;
}

export default css;
