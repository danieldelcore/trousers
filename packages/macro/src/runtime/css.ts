interface Definition {
    type: 'element' | 'modifier' | 'global' | 'theme';
    id: string;
    styles: Record<string, string>;
}

export interface CollectorReturn {
    modifier: (
        id: Definition['id'],
        styles: Definition['styles'],
    ) => CollectorReturn;
    global: (
        id: Definition['id'],
        styles: Definition['styles'],
    ) => CollectorReturn;
    theme: (
        id: Definition['id'],
        styles: Definition['styles'],
    ) => CollectorReturn;
    _get: () => Definition[];
}

function css(elementId: Definition['id'], elementStyles: Definition['styles']) {
    const styleMap: Definition[] = [
        {
            type: 'element',
            id: elementId,
            styles: elementStyles,
        },
    ];

    const self: CollectorReturn = {
        _get: () => styleMap,
        modifier: (modifierId, modifierStyles) => {
            styleMap.push({
                type: 'modifier',
                id: modifierId,
                styles: modifierStyles,
            });

            return self;
        },
        global: (globalId, globalStyles) => {
            styleMap.push({
                type: 'global',
                id: globalId,
                styles: globalStyles,
            });

            return self;
        },
        theme: (themeId, themeStyles) => {
            styleMap.push({
                type: 'theme',
                id: themeId,
                styles: themeStyles,
            });

            return self;
        },
    };

    return self;
}

export default css;
