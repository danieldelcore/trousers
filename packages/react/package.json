{
    "name": "@trousers/react",
    "version": "4.0.0",
    "main": "dist/trousers-react.cjs.js",
    "module": "dist/trousers-react.esm.js",
    "types": "dist/index.d.ts",
    "license": "MIT",
    "dependencies": {
        "@trousers/core": "3.0.2",
        "@trousers/sheet": "3.0.0"
    },
    "devDependencies": {
        "react": "^17.0.1"
    },
    "peerDependencies": {
        "react": ">=16.8.0"
    },
    "repository": "https://github.com/danieldelcore/trousers/tree/master/packages/react"
}
