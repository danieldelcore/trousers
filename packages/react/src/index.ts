// eslint-disable-next-line prettier/prettier
export type { TrousersProps } from './types';

export { default } from './jsx';
