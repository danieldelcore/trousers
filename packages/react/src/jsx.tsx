import './global-types';
import React, {
    createElement,
    hasOwnProperty,
    ElementType,
    ReactNode,
    useLayoutEffect,
    Fragment,
} from 'react';

import { TrousersProps, process, isBrowser } from '@trousers/core';
import sheet from '@trousers/sheet';

let styleSheet: ReturnType<typeof sheet> | null = null;
if (isBrowser()) {
    const headElement = document.getElementsByTagName('head')[0];
    styleSheet = sheet(headElement, 'data-trousers');
}

const jsx = <Props extends TrousersProps>(
    type: ElementType<Omit<Props, 'css'>>,
    props: Props,
    ...children: ReactNode[]
) => {
    if (props == null || !hasOwnProperty.call(props, 'css'))
        return createElement(type, props, ...children);

    //TODO: might be good to memo here
    const definitions = props
        .css!._get()
        .filter(
            ({ id, type }) =>
                type !== 'modifier' ||
                (props.hasOwnProperty(`$${id}`) && !!props[`$${id}`]),
        );

    const classes = definitions
        .map(({ className }) => className.substring(1))
        .join(' ')
        .trim();

    const cleanProps = Object.keys(props)
        .filter(key => !key.startsWith('$') && key !== 'css')
        .reduce((obj: Record<string, any>, key: string) => {
            obj[key] = props[key];
            return obj;
        }, {});

    const Element = createElement(
        type,
        // @ts-ignore
        {
            ...cleanProps,
            className: classes,
        },
        ...children,
    );

    if (!isBrowser()) {
        const styles = definitions
            .map(({ className, styles }) =>
                Object.entries(process(className, styles)).reduce(
                    (accum, [key, value]) => `${accum}${key} {${value}}`,
                    '',
                ),
            )
            .join('\n');

        return (
            <Fragment>
                <style dangerouslySetInnerHTML={{ __html: styles }} />
                {Element}
            </Fragment>
        );
    }

    // eslint-disable-next-line react-hooks/rules-of-hooks
    useLayoutEffect(() => {
        const cleanUp: string[] = [];

        definitions
            .filter(({ className }) => styleSheet && !styleSheet.has(className)) // this doesn't stop all selectors
            .forEach(({ id, className, styles, type }) =>
                Object.entries(process(className, styles)).forEach(
                    ([key, value]) => {
                        if (type === 'global') {
                            cleanUp.push(id + key);
                        }

                        return (
                            styleSheet &&
                            styleSheet.mount(
                                id + key,
                                `${key}{${value}}`,
                                type === 'global',
                            )
                        );
                    },
                ),
            );

        return () => {
            cleanUp.forEach(id => styleSheet && styleSheet.unmount(id));
        };
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [classes]);

    return Element;
};

export default jsx;
