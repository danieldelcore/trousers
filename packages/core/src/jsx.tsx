import { createElement, hasOwnProperty, ElementType, ReactNode } from 'react';

import { StyleCollector, CSSProps } from '@trousers/utils';
import useStyles from './useStyles';

import './types';

const jsx = <
    Props extends { css: StyleCollector<Theme> | CSSProps },
    Theme extends {} = {}
>(
    type: ElementType<Omit<Props, 'css'>>,
    props: Props,
    ...children: ReactNode[]
) => {
    if (props == null || !hasOwnProperty.call(props, 'css')) {
        return createElement(type, props, ...children);
    }

    const { css, ...rest } = props;

    // eslint-disable-next-line react-hooks/rules-of-hooks
    const className = useStyles<Theme>(props.css);

    return createElement(type, { ...rest, className }, ...children);
};

export default jsx;
export { CSSProps };
