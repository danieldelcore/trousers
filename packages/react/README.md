# @trousers/react

`@trousers/react` is a Trousers implementation specifically for React.

## Installation

`yarn add @trousers/react` or `npm install @trousers/react`

## API

### `jsx`

This package exports a custom `jsx` pragma which can be used to enable the CSS prop.

-   The pragma allows for HTML elements to accept a `css` prop and modifier props, denoted by the `$` symbol.
-   The pragma contains logic for mounting and processing the styles for use with React
-   Modifiers accept a boolean value. Under the hood will toggle the relevant styles on or off accordingly. (if supplied by the styles passed into the css prop).

**Example: Basic usage**

```jsx
/** @jsx jsx */
import css from '@trousers/core';
import jsx from '@trousers/react';

const styles = css('button', { backgroundColor: 'red' }).modifier('primary', {
    backgroundColor: 'blue',
});

const Button = ({ primary }) => <button css={styles} $primary={primary} />;
```
