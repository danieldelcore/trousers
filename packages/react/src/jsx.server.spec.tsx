/** @jsx jsx */
import jsx from './jsx';
import { render } from '@testing-library/react';
import css from '@trousers/core';

jest.mock('@trousers/core', () => ({
    __esModule: true,
    ...(jest.requireActual('@trousers/core') as object),
    isBrowser: () => false,
}));

describe('jsx (ssr mode)', () => {
    it('renders element styles', () => {
        const styles = css('element', {
            color: 'red',
        });

        const { container, getByTestId } = render(
            <div css={styles} data-testid="test" />,
        );
        expect(getByTestId('test').className).toMatchInlineSnapshot(
            `"element-2313942302"`,
        );
        expect(container).toMatchSnapshot();
    });

    it('renders element styles without id', () => {
        const styles = css({
            color: 'red',
        });

        const { container, getByTestId } = render(
            <div css={styles} data-testid="test" />,
        );
        expect(getByTestId('test').className).toMatchInlineSnapshot(
            `"2313942302"`,
        );
        expect(container).toMatchSnapshot();
    });

    it('does not render modifier if prop is not provided', () => {
        const styles = css('element', {
            color: 'red',
        }).modifier('primary', { color: 'blue' });

        const { container, getByTestId } = render(
            <div css={styles} data-testid="test" />,
        );
        expect(getByTestId('test').className).toMatchInlineSnapshot(
            `"element-2313942302"`,
        );
        expect(container).toMatchSnapshot();
    });

    it('does not render modifier if prop false', () => {
        const styles = css('element', {
            color: 'red',
        }).modifier('primary', { color: 'blue' });

        const { container, getByTestId } = render(
            <div css={styles} $primary={false} data-testid="test" />,
        );
        expect(getByTestId('test').className).toMatchInlineSnapshot(
            `"element-2313942302"`,
        );
        expect(container).toMatchSnapshot();
    });

    it('does not render modifier if prop undefined', () => {
        const styles = css('element', {
            color: 'red',
        }).modifier('primary', { color: 'blue' });

        const { container, getByTestId } = render(
            <div css={styles} $primary={undefined} data-testid="test" />,
        );
        expect(getByTestId('test').className).toMatchInlineSnapshot(
            `"element-2313942302"`,
        );
        expect(container).toMatchSnapshot();
    });

    it('renders element and modifier styles', () => {
        const styles = css('element', {
            color: 'red',
        }).modifier('primary', { color: 'blue' });

        const { container, getByTestId } = render(
            <div css={styles} $primary data-testid="test" />,
        );
        expect(getByTestId('test').className).toMatchInlineSnapshot(
            `"element-2313942302 element-2313942302--primary-2561700995"`,
        );
        expect(container).toMatchSnapshot();
    });

    it('renders element and modifier styles without id', () => {
        const styles = css({
            color: 'red',
        }).modifier('primary', { color: 'blue' });

        const { container, getByTestId } = render(
            <div css={styles} $primary data-testid="test" />,
        );
        expect(getByTestId('test').className).toMatchInlineSnapshot(
            `"2313942302 2313942302--primary-2561700995"`,
        );
        expect(container).toMatchSnapshot();
    });

    it('renders themed elements', () => {
        const styles = css('element', {
            color: 'var(--primary-color)',
        }).theme({ primaryColor: 'red' });

        const { container, getByTestId } = render(
            <div css={styles} data-testid="test" />,
        );
        expect(getByTestId('test').className).toMatchInlineSnapshot(
            `"element-1937164027 theme-element-4276149098"`,
        );
        expect(container).toMatchSnapshot();
    });

    it('renders themed modifiers', () => {
        const styles = css('element', {
            color: 'var(--primary-color)',
        })
            .modifier('secondary', {
                color: 'var(--secondary-color)',
            })
            .theme({
                primaryColor: 'red',
                secondaryColor: 'blue',
            });

        const { container, getByTestId } = render(
            <div css={styles} data-testid="test" />,
        );
        expect(getByTestId('test').className).toMatchInlineSnapshot(
            `"element-1937164027 theme-element-2013313543"`,
        );
        expect(container).toMatchSnapshot();
    });

    it('renders themed globals', () => {
        const styles = css('element', {
            color: 'red',
        })
            .global({
                ':root': {
                    color: 'var(--color)',
                },
            })
            .theme({ color: 'red' });

        const { container, getByTestId } = render(
            <div css={styles} data-testid="test" />,
        );
        expect(getByTestId('test').className).toMatchInlineSnapshot(
            `"element-2313942302  theme-element-2313942302"`,
        );
        expect(container).toMatchSnapshot();
    });

    it('renders single global', () => {
        const styles = css('element', {
            color: 'red',
        }).global({
            ':root': {
                boxSizing: 'border-box',
            },
        });

        const { container, getByTestId } = render(
            <div css={styles} data-testid="test" />,
        );
        expect(getByTestId('test').className).toMatchInlineSnapshot(
            `"element-2313942302"`,
        );
        expect(container).toMatchSnapshot();
    });

    it('renders multiple globals', () => {
        const styles = css('element', {})
            .global({
                ':root': {
                    boxSizing: 'border-box',
                },
            })
            .global({
                // @ts-ignore
                body: {
                    color: 'green',
                },
            });

        const { container, getByTestId } = render(
            <div css={styles} data-testid="test" />,
        );
        expect(getByTestId('test').className).toMatchInlineSnapshot(
            `"element-3938"`,
        );
        expect(container).toMatchSnapshot();
    });
});
