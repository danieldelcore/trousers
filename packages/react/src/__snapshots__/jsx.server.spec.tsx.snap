// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsx (ssr mode) does not render modifier if prop false 2`] = `
<div>
  <style>
    .element-2313942302 {color: red;}
  </style>
  <div
    class="element-2313942302"
    data-testid="test"
  />
</div>
`;

exports[`jsx (ssr mode) does not render modifier if prop is not provided 2`] = `
<div>
  <style>
    .element-2313942302 {color: red;}
  </style>
  <div
    class="element-2313942302"
    data-testid="test"
  />
</div>
`;

exports[`jsx (ssr mode) does not render modifier if prop undefined 2`] = `
<div>
  <style>
    .element-2313942302 {color: red;}
  </style>
  <div
    class="element-2313942302"
    data-testid="test"
  />
</div>
`;

exports[`jsx (ssr mode) renders element and modifier styles 2`] = `
<div>
  <style>
    .element-2313942302 {color: red;}
.element-2313942302--primary-2561700995 {color: blue;}
  </style>
  <div
    class="element-2313942302 element-2313942302--primary-2561700995"
    data-testid="test"
  />
</div>
`;

exports[`jsx (ssr mode) renders element and modifier styles without id 2`] = `
<div>
  <style>
    .2313942302 {color: red;}
.2313942302--primary-2561700995 {color: blue;}
  </style>
  <div
    class="2313942302 2313942302--primary-2561700995"
    data-testid="test"
  />
</div>
`;

exports[`jsx (ssr mode) renders element styles 2`] = `
<div>
  <style>
    .element-2313942302 {color: red;}
  </style>
  <div
    class="element-2313942302"
    data-testid="test"
  />
</div>
`;

exports[`jsx (ssr mode) renders element styles without id 2`] = `
<div>
  <style>
    .2313942302 {color: red;}
  </style>
  <div
    class="2313942302"
    data-testid="test"
  />
</div>
`;

exports[`jsx (ssr mode) renders multiple globals 2`] = `
<div>
  <style>
    
:root {box-sizing: border-box;}
 body {color: green;}
  </style>
  <div
    class="element-3938"
    data-testid="test"
  />
</div>
`;

exports[`jsx (ssr mode) renders single global 2`] = `
<div>
  <style>
    .element-2313942302 {color: red;}
:root {box-sizing: border-box;}
  </style>
  <div
    class="element-2313942302"
    data-testid="test"
  />
</div>
`;

exports[`jsx (ssr mode) renders themed elements 2`] = `
<div>
  <style>
    .element-1937164027 {color: var(--primary-color);}
.theme-element-4276149098 {--primary-color: red;}
  </style>
  <div
    class="element-1937164027 theme-element-4276149098"
    data-testid="test"
  />
</div>
`;

exports[`jsx (ssr mode) renders themed globals 2`] = `
<div>
  <style>
    .element-2313942302 {color: red;}
:root {color: var(--color);}
.theme-element-2313942302 {--color: red;}
  </style>
  <div
    class="element-2313942302  theme-element-2313942302"
    data-testid="test"
  />
</div>
`;

exports[`jsx (ssr mode) renders themed modifiers 2`] = `
<div>
  <style>
    .element-1937164027 {color: var(--primary-color);}
.theme-element-2013313543 {--primary-color: red;--secondary-color: blue;}
  </style>
  <div
    class="element-1937164027 theme-element-2013313543"
    data-testid="test"
  />
</div>
`;
